### ISSUE TITLE
RFC: Adding Image Indexes

### ISSUE THREAD

**Issue by jonjohnsonjr** [google/go-containerregistry#254]
# Adding Image Indexes

Opening a separate issue from #119 to track the implementation of just the data structures. We can use the other issue to track follow ups too.

## Background

An Image Index (AKA Manifest List (AKA Fat Manifest)) represents a set of images + some metadata.

In Docker-land, that metadata was exclusively the `Platform` (operating system + architecture), which was a required property of a Manifest List descriptor.

In OCI-land, there are annotations for arbitrary metadata, but `Platform` is still top-level (for backwards compat with docker), but not required.

## Prior Art

These tools primarily deal with the creation of image indexes.

* [estesp/manifest-tool](https://github.com/estesp/manifest-tool)
* [`docker manifest` (experimental)](https://docs.docker.com/edge/engine/reference/commandline/manifest/)

## Objective

The primary use of an Image Index is to support multi-platform images, so that will be the focus of this doc.

There are many other things that image authors can do with an image index, but supporting that should be as easy as _not_ clobbering descriptor annotations.

To start, we'll just implement a `remote.ImageIndex` primitive that exposes the information in an image index.

We can have more focused discussion on context-specific behavior once we start plumbing it through to other packages.

## Motivation

There are a few notable features that image index support will enable.

### Copying Image Indexes

One of the primary use-cases of `crane` is the ability to easily perform fast cross-registry copies.
This currently doesn't work for image indexes (all `library/` images), so we need to fix it.

### Remote inspection

A secondary use case is exploring registry contents, e.g. `crane manifest` and `crane digest`.
These are currently misleading for image indexes, since they resolve to the default amd64/linux image.

### Tag to Digest resolution

We [resolve tags to digests for Revisions](https://github.com/knative/serving/blob/b707183d8355544c3cd8e841923eb08ccb91ceba/pkg/reconciler/v1alpha1/revision/resolve.go#L37).
We'll need to support resolving image index digests instead for [windows nodes](https://github.com/knative/serving/issues/1816).

### Fat `ko`

This is the most exciting thing that image index support unblocks, and probably deserves its own RFC.

Currently, `ko` assumes a single platform base image and produces a single output image.

Once support for image indexes is implemented, `ko` could accept an image index as a base image and produce an image index as its output, cross-compiling for each `manifests` entry.

## Issues

Docker's behavior with respect to image indexes has been a frequent source of confusion.

1. When pulling an image index, Docker will "resolve" it to your current platform.
1. The digest that gets printed when pulling is for the image index, not the platform-specific manifest.

An open question for us: When do we resolve image indexes to manifests, and when do we expose the image index to users?

## Strawman

I have a [WIP branch](https://github.com/google/go-containerregistry/compare/master...jonjohnsonjr:manifest-lists).

Changes shouldn't be too huge to the core `v1` interfaces and structs.

We'll need to add some types to represent platforms and image indexes:

```go
type Index struct {
	SchemaVersion int64             `json:"schemaVersion"`
	MediaType     types.MediaType   `json:"mediaType"`
	Manifests     []Descriptor      `json:"manifests"`
	Annotations   map[string]string `json:"annotations,omitempty"`
}

type Platform struct {
	Architecture string   `json:"architecture"`
	OS           string   `json:"os"`
	OSVersion    string   `json:"os.version,omitempty"`
	OSFeatures   []string `json:"os.features,omitempty"`
	Variant      string   `json:"variant,omitempty"`
}
```

The interface I imagine for interacting with an image index. Likely I've missed something :)

```go
type ImageIndex interface {
	// MediaType of this image's manifest.
	MediaType() (types.MediaType, error)

 	// Digest returns the sha256 of this index's manifest.
	Digest() (Hash, error)

 	// Index returns this image index's manifest object.
	Manifest() (*Index, error)
	
 	// RawManifest returns the serialized bytes of Manifest().
	RawManifest() ([]byte, error)

 	// Image returns the image referenced by this index for a given platform.
	Image(Platform) (*Image, error)
}
```

And we'll need to add a `Platform` to `Descriptor`:

```go
type Descriptor struct {
	MediaType   types.MediaType   `json:"mediaType"`
	Size        int64             `json:"size"`
	Digest      Hash              `json:"digest"`
	URLs        []string          `json:"urls,omitempty"`
	Annotations map[string]string `json:"annotations,omitempty"`
	Platform    *Platform         `json:"platform,omitempty"`
}
```

**Comment by jonjohnsonjr** [comment]
Fixed by:
https://github.com/google/go-containerregistry/pull/257
https://github.com/google/go-containerregistry/pull/386
https://github.com/google/go-containerregistry/pull/387
https://github.com/google/go-containerregistry/pull/394
